RFC: Open up your enum with an unnamed variant

kupiakos opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rfcs/pull/3894)*

Enable ranges of enum discriminants to be claimed ahead of time with an unnamed enum variant. This requires all users of that enum to consider those values as valid - including within the declaring crate.

This is useful for any situation in which recompiling an enum to handle new discriminant values is infeasible, and using a newtype integer is unergonomic due to the type's intended nature as a set of enumerated values. This includes FFI, Protobuf, and embedded syscalls.

If an enum is valid for every discriminant (it has no niches), it is an _open enum_ and may be `as` cast from its explicit backing integer. For example:

```rust
#[repr(u32)]     // Fruit is represented by any `u32` due to the unnamed *open variant* below.
enum Fruit {
    Apple,       // Apple is represented with 0u32.
    Orange,      // Orange is represented with 1u32.
    Banana = 4,  // Banana is represented with 4u32.
    _ = ..,      // Unnamed variants in `Fruit` are represented with the
                 // remaining discriminants in `u32`.
}
// Using an `as` cast from `u32`, since it is an open enum.
let fruit = 3 as Fruit;

// Does not match any of the known variants.
match fruit {
    Fruit::Apple | Fruit::Orange | Fruit::Banana => unreachable!(),
    // This wildcard branch is required for all exhaustive matches of `Fruit`.
    // `fruit` preserves its value casting back to `u32`.
    _ => assert_eq!(fruit as u32, 3),
}
```

While writing this RFC I was not aware of the recent work on a [similar RFC](https://github.com/rust-lang/rfcs/pull/3803) by @madsmtm. I considered using an attribute to make an open enum before I had learned about this RFC  - I [explain](https://github.com/kupiakos/rs-rfcs/blob/unnamed-variants/text/3894-unnamed-variants.md#as-an-enum-attribute) why I chose unnamed variants instead in the Alternatives.

@rustbot label T-lang

> [!IMPORTANT]  
> When responding to RFCs, try to use inline review comments (it is possible to leave an inline review comment for the entire file at the top) instead of direct comments for normal comments and keep normal comments for procedural matters like starting FCPs.
>
> This keeps the discussion more organized.

[Rendered](https://github.com/kupiakos/rs-rfcs/blob/unnamed-variants/text/3894-unnamed-variants.md).

kupiakos committed:
Add the unnamed_variants RFC

kupiakos committed:
Add PR number, start date

kupiakos committed:
Fix typo, qualify as-cast type inference

Lokathor commented on text/3894-unnamed-variants.md:
```diff
@@ -0,0 +1,1885 @@
+## Unnamed variants
+
+An **unnamed variant** is an enum variant with `_` declared for its name. It is
+assigned to one or many **reserved discriminants**. These discriminants are
```
Compile error, same as if you'd assigned more than one name to a single value, I presume.

programmerjake replied:
> Compile error, same as if you'd assigned more than one name to a single value, I presume.

I think that would be really annoying for macros that try to create FFI-compatible enums to have to special case looking for an enum with exactly 256 variants (assuming we can ignore enums with >=65536 variants).

programmerjake replied:
> The recommendation to users who want an open enum but who also spelled out exactly 256 variants for `repr(u8)` is to do nothing.

I think that's a bad recommendation since in many cases it's not at all obvious that there are 256 variants when you're reading the code (basically any enum where you don't have the variants named after numbers), and if you documented that, documentation can be incorrect so you'd want to check. using `_ = ..` means the compiler handles it for you so all you have to do is look for `_ = ..`.

kupiakos replied:
> I think that's a bad recommendation since in many cases it's not at all obvious that there are 256 variants when you're reading the code (basically any enum where you don't have the variants named after numbers), and if you documented that, documentation can be incorrect so you'd want to check. using `_ = ..` means the compiler handles it for you so all you have to do is look for `_ = ..`.

Perhaps then we recommend  `const _: () = 0 as Enum` to  validate that an enum is open, which works regardless of the `repr`.

kupiakos replied:
> with `_ = ..` acting like `#[non_exhaustive]` like I proposed, a wildcard branch is required in downstream code. Since it's not exactly `#[non_exhaustive]`, the wildcard branch could also be required in the same crate.

Putting `non_exhaustive` on an enum that is already open is an anti-pattern - there is no way to actually add another variant that isn't a breaking change.

kupiakos replied:
The presence of `_ = ..` always confirms that an enum is open, but the absence of such does not imply it is closed. An enum can also be open with multiple unnamed ranges, such as:

```rust
#[repr(u8)]
enum Foo {
    X = 0,
    _ = 1..100,
    Y = 100,
    _ = 101..
}
```

So "just check for `_ = ..`" is not sufficient to know if an enum is open.

kupiakos replied:
I'm not opposed to allowing this behind a warning, though.

If it's allowed, then we must define the behavior when an empty range (`start >= end`) is provided for a discriminant range. That was a situation I didn't want to accept as that seemed hazardous. Currently that's rejected as an extension of the rule that an unnamed variant must reserve at least one discriminant (it must have an effect).

tmandry replied:
Downgrading to a warning for the unnecessary `_ = ..` case makes sense to accommodate macros.

I think there should always be a warning if a non-`..` range overlaps with another variant.

That said, procedurally, the RFC only needs to specify what is and is not an error. It's still a good idea to say what can/should be a lint, but generally RFCs do not _require_ the existence of a lint, since lints can be created and changed without an RFC.

(edit for clarification: It is common for the lang team to hold up stabilization until certain lints are implemented, and the RFC stating that a lint _should_ exist is usually enough to motivate this.)

kupiakos committed:
Clarify deserialization of bytes in embedded

There is no reason for these bytes to be in flash memory - it is merely
one of the applicable use cases.

kupiakos committed:
Add `enum` to feature name, clarify zero-copy further

unnamed_variants is an unambiguous feature name that describes what the
author is now able to spell. However, it takes more mental effort to
remember what a variant is, while "enum variant" is immediately clear.

This also clarifies the zero-copy deserialization motivation language
further.

kupiakos commented on text/3894-unnamed-variants.md:
```diff
@@ -0,0 +1,1884 @@
+- Feature Name: (`unnamed_variants`)
```
Agreed! Done.

I don't mind casually calling this the "open enum RFC" - it's catchy! However, I do think that the feature name itself should describe what the author can spell when enabled.

kupiakos replied:
I realized I hadn't updated the file name, just the feature name.

A) Is that important to rename before merging?
B) I'd prefer to keep the file name as-is for at least a little bit, as I've sent out a link to folks pointing directly to that file in my repo. Those would 404 if I renamed it at this point, so I'd prefer to have that be the last task before merging.

kupiakos committed:
Increase heading levels

See f17e862 for reasoning.

kupiakos committed:
Allow no-effect unnamed variants as a lint, add alternative

The `taken_discrimimant_ranges` and `empty_discriminant_ranges` lints
catch the situation in which an unnamed variant has no effect. The
latter is much more dangerous than the former, so it is
`deny`-by-default. However, both should be allowed for certain codegen
and macro cases as described.

This also moves around and expands some language regarding `derive`
difficulty in open-enum, as well as add a suggested alternative.

kupiakos committed:
Extend drawbacks of open enum via attribute

Based on this comment by @dhardy:
https://github.com/rust-lang/rfcs/pull/3803#issuecomment-325722825

kupiakos committed:
Fix typo in empty discriminant ranges

kupiakos committed:
Fix empty_discriminant_ranges and C# typos

kupiakos commented on text/3894-unnamed-variants.md:
```diff
@@ -0,0 +1,1961 @@
+
+- C++'s [scoped enumerations][cpp-scoped-enums] and C enums are both open
+  enums.
+- C## uses [open enums][cs-open-enums], with a [proposal][cs-closed-enums] to
```
Ah, I thought I'd caught everything, thanks. I took the opportunity to give it the official name as well 🎶

kupiakos committed:
Fix implicit/explicit swap

scottmcm commented on text/3894-unnamed-variants.md:
```diff
@@ -0,0 +1,1963 @@
+  - The unstable [`non_exhaustive_omitted_patterns`] lint has no easy way to
+    work with this enum-alike, even though treating it like a `non_exhaustive`
+    enum would be more helpful.
+- Generated rustdoc is less clear (the pseudo-enum is grouped with `struct`s).
```
TBH, I've long found this distinction in rustdoc unhelpful *always*.  If it's an opaque type, for example, what do I care if it's a `struct` or a `union`?  Not to mention that I'd rather have `Statement` and `StatementKind` next to each other on <https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/> rather than multiple pages apart.

I'd absolutely 👍 an RFC to change rustdoc to group things in a module by *namespace* instead of *item kind*, but I don't think the current behaviour is a reason to do a *language* change.

scottmcm replied:
(Using colour or an icon or something for `struct` vs `enum` to help give my brain an extra cue as I scan?  Sure, great!  I just don't think it makes sense as a top-level split.)

kupiakos replied:
I would love to have more control over rustdoc presentation and organization of types as well, in general. I consider the accurate grouping in rustdoc for an open enum as more removing a papercut than a full motivation for the feature.

kupiakos replied:
Would you like me to change the RFC text to reflect that a hypothetical future change to rustdoc could negate this particular issue?

programmerjake commented on text/3894-unnamed-variants.md:
```diff
@@ -0,0 +1,1963 @@
+- Crates must be recompiled to use new enum variants.
+- It affects _only_ foreign crates.
+
+By contrast, an unnamed variant affects API _and_ ABI semver compatibility:
```
having it affect API is necessary to handle stuff like:
```rust
#[repr(u8)]
enum E {
    A = 0,
    B = 1,
    _ = ..,
}

match 4u8 as E {
    E::A => String::from("A"),
    E::B => String::from("B"),
    // no other match arms, what does this do?
    // the only reasonable option is a compile error.
    // that's why you need a wildcard match arm
}
```

kmicklas replied:
Yeah a lint might be sufficient. Alternatively I guess there could be special syntax to match unnamed variants, distinct from the generic wildcard which matches named variants.

programmerjake replied:
> I do think there should be lint that triggers if you don't explicitly handle one of the known variants.

there is a nightly lint [`non_exhaustive_omitted_patterns`](https://github.com/rust-lang/rust/issues/89554) which you can change to warn when `_` matches any enum variants in a `#[non_exhaustive]` enum.

kupiakos replied:
> there is a nightly lint [`non_exhaustive_omitted_patterns`](https://github.com/rust-lang/rust/issues/89554)

Indeed, that's the desired lint. The interaction of this lint with unnamed variants [is described in the RFC](https://github.com/kupiakos/rs-rfcs/blob/unnamed-variants/text/3894-unnamed-variants.md#forgot-to-mention-a-named-variant) and allows a maintainer to ensure that a `match` doesn't forget to list a named variant. As mentioned there, it could be renamed before stabilization if the name mismatch seems confusing. I don't think it makes sense to separate into a different lint.

```rust
#[repr(u32)]
enum Bar {
    A,
    B,
    _ = ..,
}
let b = Bar::A;

// warning: some variants are not matched explicitly
//          pattern `Bar::B` not covered
// help: ensure that all variants are matched explicitly by adding the
//       suggested match arms
// note: the matched value is of type `Bar` and the
//       `non_exhaustive_omitted_patterns` attribute was found
#[warn(non_exhaustive_omitted_patterns)]
let name = match b {
    Bar::A => "A",
    _ => "unknown",
};
```

> I guess there could be special syntax to match unnamed variants, distinct from the generic wildcard which matches named variants

It's quite intentional that unnamed variants make naming a previously unnamed variant a non-breaking API change, which a special syntax for matching on named/unnamed variants would violate. However, I do think it's occasionally useful to be able to query whether a value is a named variant, and so this operation is trivial to generate [with a macro](https://github.com/kupiakos/rs-rfcs/blob/unnamed-variants/text/3894-unnamed-variants.md#an-other-variant-carries-unknown-discriminants-like-a-tuple-variant). It's equivalent to a `matches!` on a `|` pattern that lists every named variant. Macros are allowed to cause a non-breaking API change to a type's structure yield a breaking API change; see `derive(PartialEq)` and anything else that works per-`struct`-field.

The syntax would thus be something like:

```rust
#[repr(u32)]
#[derive(IsNamedVariant)]
enum Bar {
    A,
    B,
    _ = ..,
}
let b = Bar::A;

fn known_value(b: Bar) -> Option<u32> {
    x.is_named_variant().then(|| x as u32)
}

assert_eq!(known_value(0 as Bar), Some(0));

// If you add `C` with discriminant 2, this panics.
assert_eq!(known_value(2 as Bar), None);
```

I think it's best to leave this functionality to 3rd-party macros.

> I can't see any reason it would be desirable to bundle in the API part to this feature when that's what `#[non_exhaustive]` already does

I describe my reasons for having this _replace_ `non_exhaustive` rather than require it to use unnamed enum variants [here](https://github.com/kupiakos/rs-rfcs/blob/unnamed-variants/text/3894-unnamed-variants.md#require-non_exhaustive-dont-forbid-it).

kupiakos committed:
Add rationale rejecting implicit discriminants and no-repr

Based on
[this comment](https://github.com/rust-lang/rfcs/pull/3894#issuecomment-3682414869).

kupiakos commented on text/3894-unnamed-variants.md:
```diff
@@ -0,0 +1,1988 @@
+  variant by changing the `repr`. This is non-obvious and can be avoided by
+  forbidding `non_exhaustive` when a valid unnamed variant exists.
+
+### Allow an implicit discriminant expression for unnamed variants
```
> Nice RFC!

Thanks @tmvkrpxl0  😄 

Please reply in this inline comment to keep the main thread cleaner, or resolve the thread.

> * `_` variant without pattern assigned to it should be same as `_ = ..`

I think that's ambiguous -  why wouldn't it be equivalent to `_ = <previous variant + 1`? Unnamed enum variants are unusual enough that I think making it clear that discriminants are involved should be mandatory.

> * If it's used without `#[repr]`, it should behave same as `#[non_exhaustive]`.

I think this adds more confusion than improves ergonomics. I don't think it makes as much sense to talk about assigning "the rest" of the discriminants to an enum without a `repr` as `_ = ..`  says.

>   * Given that it isn't breaking change to replace `non_exhaustive`, it already work as its replacement. If it cannot guarantee ABI compatibility due to missing `#[repr]`, it should instead ensure API compatibility only, effectively replacing it.

It doesn't replace `non_exhaustive`, because it can also be applied to `struct`s and `enum` variants. The [unnamed fields](https://github.com/rust-lang/rfcs/pull/2102) RFC also requires `repr(C)` in order to be used since it's about ABI.

> * There should be way to match for only known enum variants or only unknowns
>   
>   * I can see someone printing warning when enum value isn't known.

See [this reply](https://github.com/rust-lang/rfcs/pull/3894#discussion_r2646519115) - there is both an existing nightly lint to warn about unlisted named variants, and getting a `bool` for whether a variant is named (for a `match` guard) can be trivially done with a macro.

scottmcm commented:
> Consider this enum that we want to be made open with an [`open_enum`](https://docs.rs/open-enum/latest/open_enum/attr.open_enum.html) attribute:

To me this feels like a fixation on a particular UI for it.  You could always have an `open_enum! { ... }` non-attribute to pass everything to.  Or have both the `enum Foo` and a generated `struct FooOpen` that's also available via `Foo::Open` and has a `Foo: TryFrom<FooOpen>` that's also generated.  That'd be useful anyway for cases that *don't* want to always just pass things along but want to handle the things it actually knows about.

joshtriplett commented on text/3894-unnamed-variants.md:
```diff
@@ -0,0 +1,1988 @@
+  callsite. Thus it may be worth adding to Rust even if `.0` isn't.
+- When should one prefer the constructor over the `as` cast? Always?
+
+#### Discriminant field access
```
There have been a few other proposals for discriminant access syntax. @scottmcm has made a couple, in particular. I'd want to see those captured as alternatives here. `.0` is cute but *entirely* too clever for its own good. :)

RalfJung replied:
> though I think there were better ways of pointing it out than implying I hadn't bothered to search.

This is fair, sorry for that. I mostly mentioned "quick search" to avoid any impression that this might be an exhaustive list.

kupiakos replied:
This has been moved to an [above section under Alternatives](https://github.com/kupiakos/rs-rfcs/blob/unnamed-variants/text/3894-unnamed-variants.md#forward-compatibility-with-newtype-structs)

joshtriplett commented on text/3894-unnamed-variants.md:
```diff
@@ -0,0 +1,1988 @@
+
+None.
+
+## Future possibilities
```
I've seen proposals to have the catch-all variant support having a field that gets the discriminant, like `#[bikeshed_catchall_variant] UnknownVariant(uN)`. I think that's worth recording as an alternative, though I think it *isn't* preferable if we have a uniform syntax for both accessing discriminants and constructing enums with arbitrary discriminants.

kupiakos commented on text/3894-unnamed-variants.md:
```diff
@@ -0,0 +1,1988 @@
+    Green = 1,
+    // Must specify a non-overlapping range,
+    // including if `..` is the discriminant.
+    Unknown = 2..=50,
```
1. The reason why this Future possibility excludes named variants from overlapping other variants is because it resolves most of the issues described [in this comment](https://github.com/rust-lang/rfcs/pull/3894#discussion_r2684520800) and [section of the RFC](https://github.com/kupiakos/rs-rfcs/blob/unnamed-variants/text/3894-unnamed-variants.md#discriminant-ranges-for-named-variants-instead-of-unnamed-variants).
1. This Future possibility is specifically about _named_ variants with a range of discriminants.
1. This RFC **does** [describe the behavior](https://github.com/kupiakos/rs-rfcs/blob/unnamed-variants/text/3894-unnamed-variants.md#unnamed-variants) of `2..` and `..=50` for _unnamed_ variants.
1. I agree that `..` is by far the most important case, though it's not significantly more work to support smaller ranges. As you say, most of the work is figuring out [when you should lint](https://github.com/kupiakos/rs-rfcs/blob/unnamed-variants/text/3894-unnamed-variants.md#gap-of-length-one-caused-by-an-exclusive-range) on possibly-confusing cases.

kupiakos committed:
Clarify Swift prior art

kupiakos committed:
Reword some alternative discussion

kupiakos committed:
Use 'claim' instead of 'reserve' for discriminants

The term 'reserve' can be confusing to those familiar with networking
protocols, as it often means "was used in the past and can no longer be
used any more" as in Protobuf's `reserved` keyword.

This commit switches to using 'claim' instead, and avoids giving a
particular name to the set of discriminants that are solely assigned to
unnamed variants.

kupiakos committed:
Mention default discriminant attribute for Other-variant alt

kupiakos committed:
Clean up unnamed variant vs. declaration

Distinguish between unnamed variant declarations (the actual syntax),
and the individual unnamed variants that each discriminant represents
in the declared range.

Also clean up some other language.

madsmtm commented on text/3894-unnamed-variants.md:
```diff
@@ -0,0 +1,2030 @@
+[bytemuck-checkedbitpattern]: https://docs.rs/bytemuck/latest/bytemuck/checked/trait.CheckedBitPattern.html
+[zerocopy-frombytes-derive]: https://docs.rs/zerocopy/0.6.1/zerocopy/derive.FromBytes.html
+
+### Restricted range integers
```
I don't believe this is a very strong use-case (niche-types are the better solution for it), but it feels larger when reading the RFC because of the complex code example.

Maybe we could move the code into a playground link instead?

kupiakos committed:
Clarify explicit `repr` is required for discriminant field access

It's too implicit if the section is taken out of context.

kupiakos committed:
Link other proposals for discriminant field access

kupiakos committed:
Describe syntax of enum discriminant alternatives

Also use a more stable link for unmerged/proposed RFCs.

kupiakos committed:
Clarify "No field data" section

kupiakos committed:
Rewrite the Compatibility section and clarify terminology

Rewritten to clarify the valid conversion from `struct` and describe
CFI.

s/forwards/forward/ compatibility
s/backwards/backward/ compatibility
s/foreign/downstream/

kupiakos committed:
Add a syntax "sugar" section

The "sugar" is in quotes due to this being able to desugar to
enums with an impossibly large number of variants.

kupiakos committed:
Re-introduce `..` at the end as an alternative

kupiakos committed:
Clarify some alternatives, move tuple discriminant access there

kupiakos committed:
Add `repr(transparent, Int)` alternative for ABI compat

kupiakos committed:
Various nits

Jules-Bertholet commented:
An example use-case for this that has nothing to do with C: the standard library's [`Ipv6MulticastScope`](https://doc.rust-lang.org/nightly/std/net/enum.Ipv6MulticastScope.html) enums has two variants that are perma-unstable and `doc(hidden)` essentially as a hack to emulate this feature.

kupiakos committed:
Correct current usage of enums in Rust for Linux

kupiakos committed:
Improve CFI section

kupiakos committed:
Clean up repr(C) behavior language

kupiakos committed:
Improve CFI section further

kupiakos committed:
Replace `repr(transparent, Int)` enum with better alts

kupiakos committed:
Clean up forward compat with newtype struct section

kupiakos committed:
Various nits

kupiakos committed:
Add future work section for pattern type casts

kupiakos committed:
Use "underlying" instead of "backing" integer

While "representing" may be better, "underlying integer" matches the
language of the C standard.

kupiakos committed:
Lint instead of reject non_exhaustive and unnamed variants

`non_exhaustive` on an enum with unnamed variants is instead considered
"unused", which is a warning by default.

kupiakos committed:
Mention Ipv6MulticastScope in Motivation

kupiakos committed:
Use "detect" rather than "produce" for lints

This mirrors the language used by official docs.

kupiakos committed:
Add rust-lang/rust tracking issue for unnamed variants

kupiakos committed:
Use `-` not `_` for lint reference

Again, to be consistent with official docs.

kupiakos committed:
Update guide-level `non_exhaustive` section

Darksonn commented on text/3894-unnamed-variants.md:
```diff
@@ -0,0 +1,2404 @@
+  - For `repr(C)`, the `Int` must be ABI compatible with the target's chosen
+    integer type for a C `enum` with an equivalent definition. This is usually
+    `core::ffi::c_int`.
+  - This defines a `repr(Int)` `enum` as having the same ABI as `Int`. See
+    [Control Flow Integrity](#control-flow-integrity) for ABI caveats.
```
The nomicon currently says:

> Note that the function call ABI for these types (`repr(Int)` enums) is still in general unspecified, except that across extern "C" calls they are ABI-compatible with C enums of the same sign and size.
> 
> https://doc.rust-lang.org/nomicon/other-reprs.html#repru-repri

But C enums do not necessarily have the same ABI as the underlying integer type as seen in the CFI section, so these two statements are inconsistent.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3894/208151010f8cba1d69c354eb79dcbb6ee5b25dc4..1b2c28cf50a6ba05545189b5e8bf17b96dba6809)*

Jules-Bertholet replied:
Everything about CFI is still unstable, so until the lang team signs off on it, the non-CFI rules continue to apply.

RalfJung replied:
Even so, we shouldn't add *new* CFI-incorrect ABI guarantees. And currently we don't  make any guarantee of Rust-to-Rust ABI compatibility between a repr(Int) enum and any integer type.
